File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/BeginInvokePairedWithEndInvoke.cs
using System;
using System.Threading;

namespace Tests.Diagnostics
{
    public delegate void AsyncMethodCaller(String name);

    class Program
    {
        static void Main(string[] args)
        {
            BeginInvokeAndEndInvokeOnDelegateWithoutCallback();
        }

        private static void BeginInvokeOnDelegateWithoutCallback()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            caller.BeginInvoke("delegate", /* callback */null, /* state */ null); // Noncompliant {{Pair this "BeginInvoke" with an "EndInvoke".}}
            // https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/calling-synchronous-methods-asynchronously
            // «Important: No matter which technique you use, always call EndInvoke to complete your asynchronous call.»
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithoutCallback()
        {
            Console.WriteLine("BeginInvokeAndEndInvokeOnDelegateWithoutCallback");
            var caller = new AsyncMethodCaller(AsyncMethod);
            IAsyncResult result = caller.BeginInvoke("delegate", /* callback */null, /* state */ null); // Compliant
            caller.EndInvoke(result);
        }

        private static void BeginInvokeOnDelegateWithLambdaCallback1()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            caller.BeginInvoke("delegate", result => { }, null); // Noncompliant
        }

        private static void BeginInvokeOnDelegateWithLambdaCallback2()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            var callback = new AsyncCallback(result => { });
            caller.BeginInvoke("delegate",  callback, null); // Noncompliant
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithLambdaCallback1()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            caller.BeginInvoke(name: "delegate", @object: null, callback: result => caller.EndInvoke(result)); // Compliant
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithLambdaCallback2()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            var callback = new AsyncCallback(result => caller.EndInvoke(result));
            caller.BeginInvoke("delegate", callback, null); // Compliant, EndInvoke is called by callback
        }

        private static void BeginInvokeOnDelegateWithDelegateCallback()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            caller.BeginInvoke("delegate",  delegate(IAsyncResult result) { Console.WriteLine(); }, null); // Noncompliant
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithDelegateCallback()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            caller.BeginInvoke("delegate",  delegate(IAsyncResult result) { caller.EndInvoke(result); }, null); // Compliant
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithVariableCallback()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            AsyncCallback callback = result => { caller.EndInvoke(result); };
            caller.BeginInvoke("delegate",  callback, null); // Compliant
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithStaticCallback1()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            var callback = new AsyncCallback(StaticCallEndInvoke);
            caller.BeginInvoke("delegate",  callback, null); // Compliant, EndInvoke is called by callback and it's StaticCallEndInvoke
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithStaticCallback2()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            caller.BeginInvoke("delegate",  new AsyncCallback(StaticDoNothing), null); // Noncompliant
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithStaticCallback3()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            var callback = new AsyncCallback(StaticDoNothing);
            caller.BeginInvoke("delegate",  callback, null); // Noncompliant
        }

        private static void BeginInvokeOnDelegateWithCallbackAssignment()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            AsyncCallback callback;
            callback = new AsyncCallback(StaticDoNothing);
            caller.BeginInvoke("delegate",  callback, null); // false-negative, we only look at the variable initialization and not at all its assignments
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithWrapperCallback1()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            var wrapper = new CallerWrapper(caller);
            var callback = new AsyncCallback(wrapper.CallEndInvoke);
            caller.BeginInvoke("delegate",  callback, null); // Compliant, EndInvoke is called by wrapper.CallEndInvoke
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithWrapperCallback2()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            var wrapper = new CallerWrapper(caller);
            var callback = new AsyncCallback(wrapper.DoNothing);
            caller.BeginInvoke("delegate",  callback, null); // Noncompliant
        }

        private static void BeginInvokeAndEndInvokeOnDelegateWithNonExistingCallback()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            var callback = new AsyncCallback(Console.WriteLine);
            caller.BeginInvoke("delegate",  callback, null); // Compliant
        }

        private static void BeginInvokeOnAnyClassButDelegate()
        {
            var notADelegate = new AnyClassWithOptionalEndInvoke();
            var result = notADelegate.BeginInvoke(new AsyncMethodCaller(AsyncMethod));
            // Compliant, NotADelegate class declared below does not required a call to EndInvoke
            // Same as System.Windows.Forms.Control see
            // https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.begininvoke?view=netframework-4.6
            // «You can call EndInvoke to retrieve the return value from the delegate, if necessary, but this is not required.»
        }

        private static void AsyncMethod(string msg)
        {
            Console.WriteLine($"AsyncMethod: {msg}");
        }

        private static void StaticCallEndInvoke(IAsyncResult result)
        {
            AsyncMethodCaller caller = (AsyncMethodCaller) ((AsyncResult) result).AsyncDelegate;
            caller.EndInvoke(result);
        }

        private static void StaticDoNothing(IAsyncResult result)
        {
        }

        private static void EndInvokeOfDifferentAction()
        {
            Action a = () => { };
            Action b = () => { };
            AsyncCallback callback = b.EndInvoke;
            a.BeginInvoke(callback, null); // FN
        }

        private static void BeginInvokeAndEndInvokeOnDifferentDelegateWithVariableCallback()
        {
            var caller = new AsyncMethodCaller(AsyncMethod);
            var caller2 = new AsyncMethodCaller(AsyncMethod);
            AsyncCallback callback = result => { caller2.EndInvoke(result); };
            caller.BeginInvoke("delegate", callback, null); // FN
        }

        public class CallerWrapper
        {
            private AsyncMethodCaller caller;

            public  CallerWrapper(AsyncMethodCaller caller)
            {
                this.caller = caller;
            }

            public void CallEndInvoke(IAsyncResult result)
            {
                caller.EndInvoke(result);
            }

            public void DoNothing(IAsyncResult result)
            {
            }

        }

        public class AnyClassWithOptionalEndInvoke
        {

            public IAsyncResult BeginInvoke(AsyncMethodCaller method)
            {
                return method.BeginInvoke("NotADelegate", result => { method.EndInvoke(result); }, null);
            }

            // It's not required to call EndInvoke after BeginInvoke on this class
            public object EndInvoke(IAsyncResult asyncResult)
            {
                return null;
            }

        }

        // Coverage
        public class Foo
        {
            public static AsyncMethodCaller caller = new AsyncMethodCaller(AsyncMethod);
            public IAsyncResult result = caller.BeginInvoke("Foo", null, null); // Noncompliant
            public Action action1 = delegate
            {
                caller.BeginInvoke("Foo", null, null); // Noncompliant
            };
            public Action action2 = () =>
            {
                caller.BeginInvoke("Foo", null, null); // Noncompliant
            };
            public Action<string> action3 = name =>
            {
                caller.BeginInvoke(name, null, null); // Noncompliant
            };

            public int prop
            {
                get
                {
                    caller.BeginInvoke("prop", null, null); // Noncompliant
                    return 0;
                }
            }

            public static implicit operator int(Foo f)
            {
                caller.BeginInvoke("prop", null, null); // Noncompliant
                return 0;
            }

            public static Foo operator +(Foo b, Foo c)
            {
                caller.BeginInvoke("prop", null, null); // Noncompliant
                return new Foo();
            }

            static Foo()
            {
                caller.BeginInvoke("Foo", null, null); // Noncompliant
            }

            public Foo()
            {
                caller.BeginInvoke("Foo", null, null); // Noncompliant
            }

            public void Compliant()
            {
                IAsyncResult result = caller.BeginInvoke("method", null, null); // Compliant
                caller.EndInvoke(result);
            }

            public void Bar()
            {
                caller.BeginInvoke("Foo", null, null); // Noncompliant
            }

            ~Foo()
            {
                caller.BeginInvoke("Foo", null, null); // Noncompliant
            }

            private void InvokeSomethingElse()
            {
                var BeginInvoke = "MemberBinding";
                var EndInvoke = "MemberBinding";
                BeginInvoke.ToString();
                EndInvoke.ToString();
            }

            public struct FooStruct
            {
                public string field;

                public FooStruct(string field)
                {
                    this.field = field;
                    caller.BeginInvoke("FooStruct", null, null); // Noncompliant
                }
            }

            public void Container()
            {
                IAsyncResult result = BeginInvokeHiddenInALocalFunction();
                caller.EndInvoke(result);

                IAsyncResult BeginInvokeHiddenInALocalFunction()
                {
                    return caller.BeginInvoke("method", null, null); // Noncompliant
                }
            }
        }

        public class FakeProperty
        {
            private AsyncMethodCaller caller;

            public int Prop
            {
                get
                {
                    caller.BeginInvoke("prop", null, null); // FN, detection thinks that "EndInvoke" in setter is valid pair for this
                    return 0;
                }
                set
                {
                    caller.EndInvoke(null);
                }
            }
        }
    }

    // Dummy implementation of IAsyncResult compatible with both .Net Framework and .Net Core
    internal class AsyncResult : IAsyncResult
    {
        public object AsyncState { get; }

        public WaitHandle AsyncWaitHandle { get; }

        public bool CompletedSynchronously { get; }

        public bool IsCompleted { get; }

        public virtual object AsyncDelegate { get; }
    }

    class ReproEndinvokeDelegate
    {
        public void BeginInvokeWithEndinvokeDelegate()
        {
            Action a = () => { };
            a.BeginInvoke(a.EndInvoke, null); // Compliant
        }

        public void AsyncCallbackLocalVariable()
        {
            Action a = () => { };
            AsyncCallback callback = a.EndInvoke;
            a.BeginInvoke(callback, null); // Compliant
        }

        public void AsyncCallbackLocalVariableNotEndinvokeAccess()
        {
            Action a = () => { };
            AsyncCallback callback = a.NotEndInvoke; // Error [CS1061]
            a.BeginInvoke(callback, null); // Noncompliant
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/BeginInvokePairedWithEndInvoke.vb
Imports System
Imports System.Collections.Generic
Imports System.Threading

Public Delegate Sub AsyncMethodCaller(Name As String)

Module Common

    Public Sub AsyncMethod(Msg As String)
        Console.WriteLine($"AsyncMethod: {Msg}")
    End Sub

    Public Sub Main()
        BeginInvokeAndEndInvokeOnDelegateWithoutCallback()
    End Sub

    Private Sub BeginInvokeOnDelegateWithoutCallbackSub()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Caller.BeginInvoke("delegate", Nothing, Nothing) ' Noncompliant {{Pair this "BeginInvoke" with an "EndInvoke".}}
        ' https:'docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/calling-synchronous-methods-asynchronously
        ' «Important: No matter which technique you use, always call EndInvoke to complete your asynchronous call.»
    End Sub

    Private Function BeginInvokeOnDelegateWithoutCallbackFunction() As String
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Caller.BeginInvoke("delegate", Nothing, Nothing) ' Noncompliant {{Pair this "BeginInvoke" with an "EndInvoke".}}
        ' https:'docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/calling-synchronous-methods-asynchronously
        ' «Important: No matter which technique you use, always call EndInvoke to complete your asynchronous call.»
    End Function

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithoutCallback()
        Console.WriteLine("BeginInvokeAndEndInvokeOnDelegateWithoutCallback")
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Result As IAsyncResult = Caller.BeginInvoke("delegate", Nothing, Nothing) ' Compliant
        Caller.EndInvoke(Result)
    End Sub

    Private Sub BeginInvokeOnDelegateWithLambdaCallback1()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Caller.BeginInvoke("delegate", Sub(Result)          ' Noncompliant
                                       End Sub, Nothing)
    End Sub

    Private Sub BeginInvokeOnDelegateWithLambdaCallback2()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Callback As New AsyncCallback(Sub(Result)
                                          End Sub)
        Caller.BeginInvoke("delegate", Callback, Nothing) ' Noncompliant
    End Sub

    Private Sub BeginInvokeOnDelegateWithLambdaCallback3()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Callback As AsyncCallback = New AsyncCallback(Sub(Result)
                                                          End Sub)
        Caller.BeginInvoke("delegate", Callback, Nothing) ' Noncompliant
    End Sub

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithLambdaCallback1()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Caller.BeginInvoke(Name:="delegate", DelegateAsyncState:=Nothing, DelegateCallback:=Sub(Result) Caller.EndInvoke(Result))   ' Compliant
        Caller.BeginInvoke(Name:="delegate", DelegateAsyncState:=Nothing, DelegateCallback:=Sub(Result)                             ' Compliant
                                                                                                Caller.EndInvoke(Result)
                                                                                            End Sub)
    End Sub

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithLambdaCallback2()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Callback As New AsyncCallback(Sub(Result) Caller.EndInvoke(Result))
        Caller.BeginInvoke("delegate", Callback, Nothing) ' Compliant, EndInvoke is called by Callback
    End Sub

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithVariableCallback()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Callback As AsyncCallback = Sub(Result) Caller.EndInvoke(Result)
        Caller.BeginInvoke("delegate", Callback, Nothing) ' Compliant
    End Sub

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithStaticCallback1()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Callback As New AsyncCallback(AddressOf SharedCallEndInvoke)
        Caller.BeginInvoke("delegate", Callback, Nothing) ' Compliant, EndInvoke is called by SharedCallEndInvoke
    End Sub

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithStaticCallback2()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Caller.BeginInvoke("delegate", New AsyncCallback(AddressOf SharedDoNothing), Nothing) ' Noncompliant
    End Sub

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithStaticCallback3()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Callback As New AsyncCallback(AddressOf SharedDoNothing)
        Caller.BeginInvoke("delegate", Callback, Nothing) ' Noncompliant
    End Sub

    Private Sub BeginInvokeOnDelegateWithCallbackAssignment()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Callback As AsyncCallback
        Callback = New AsyncCallback(AddressOf SharedDoNothing)
        Caller.BeginInvoke("delegate", Callback, Nothing) ' false-negative, we only look at the variable initialization and not at all its assignments
    End Sub

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithWrapperCallback1()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Wrapper As New CallerWrapper(Caller)
        Dim Callback As New AsyncCallback(AddressOf Wrapper.CallEndInvoke)
        Caller.BeginInvoke("delegate", Callback, Nothing) ' Compliant, EndInvoke is called by wrapper.CallEndInvoke
    End Sub

    Private Sub BeginInvokeAndEndInvokeOnDelegateWithWrapperCallback2()
        Dim Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
        Dim Wrapper As New CallerWrapper(Caller)
        Dim Callback As New AsyncCallback(AddressOf Wrapper.DoNothing)
        Caller.BeginInvoke("delegate", Callback, Nothing) ' Noncompliant
    End Sub

    Private Sub BeginInvokeOnAnyClassButDelegate()
        Dim NotADelegate As New AnyClassWithOptionalEndInvoke()
        Dim Result As IAsyncResult = NotADelegate.BeginInvoke(New AsyncMethodCaller(AddressOf AsyncMethod))
        ' Compliant, NotADelegate class declared below does not required a call to EndInvoke
        ' Same as System.Windows.Forms.Control see
        ' https:'docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.begininvoke?view=netframework-4.6
        ' «You can call EndInvoke to retrieve the return value from the delegate, if necessary, but this is not required.»
    End Sub

    Private Sub SharedCallEndInvoke(Result As IAsyncResult)
        Dim Caller As AsyncMethodCaller = DirectCast(DirectCast(Result, AsyncResult).AsyncDelegate, AsyncMethodCaller)
        Caller.EndInvoke(Result)
    End Sub

    Private Sub SharedDoNothing(Result As IAsyncResult)
    End Sub

    Public Sub NullConditionalIndexing(List As List(Of Integer))
        Dim Value As Integer = List?(0)
    End Sub

End Module

Public Class CallerWrapper

    Private Caller As AsyncMethodCaller

    Public Sub New(Caller As AsyncMethodCaller)
        Me.Caller = Caller
    End Sub

    Public Sub CallEndInvoke(Result As IAsyncResult)
        Caller.EndInvoke(Result)
    End Sub

    Public Sub DoNothing(Result As IAsyncResult)
    End Sub

End Class

Public Class AnyClassWithOptionalEndInvoke

    Public Function BeginInvoke(Method As AsyncMethodCaller) As IAsyncResult
        Return Method.BeginInvoke("NotADelegate", Sub(ar) Method.EndInvoke(ar), Nothing)
    End Function

    ' It's not required to call EndInvoke after BeginInvoke on this class
    Public Function EndInvoke(asyncResult As IAsyncResult) As Object
        Return Nothing
    End Function

End Class

Public Module CommonMod

    Public Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
    Public Result As IAsyncResult = Caller.BeginInvoke("Foo", Nothing, Nothing)     ' Noncompliant
    Public Property AutoProperty As IAsyncResult = Caller.BeginInvoke("Foo", Nothing, Nothing)     ' Noncompliant

    Public Class Fake

        Public Sub Misleading()
            Caller.EndInvoke(Nothing)
        End Sub

    End Class

End Module

Public Class ForCoverage

    Public Shared Caller As New AsyncMethodCaller(AddressOf AsyncMethod)
    Public Result As IAsyncResult = Caller.BeginInvoke("Foo", Nothing, Nothing)     ' Noncompliant
    Public Property AutoProperty As IAsyncResult = Caller.BeginInvoke("Foo", Nothing, Nothing)     ' Noncompliant

    Public Action1 As Action = Sub()
                                   Caller.BeginInvoke("Foo", Nothing, Nothing)      ' Noncompliant
                               End Sub

    Public Action2 As Action = Sub() Caller.BeginInvoke("Foo", Nothing, Nothing)    ' Noncompliant

    Public Action3 As Action(Of String) = Function(Name) Caller.BeginInvoke(Name, Nothing, Nothing) ' Noncompliant

    Public Action4 As Action(Of String) = Function(Name)
                                              Return Caller.BeginInvoke(Name, Nothing, Nothing) ' Noncompliant
                                          End Function

    Public ReadOnly Property Prop As Integer
        Get
            Caller.BeginInvoke("prop", Nothing, Nothing) ' Noncompliant
        End Get
    End Property

    Public Property PropFake As Integer
        Get
            Caller.BeginInvoke("prop", Nothing, Nothing) ' Noncompliant
        End Get
        Set(value As Integer)
            Caller.EndInvoke(Nothing)
        End Set
    End Property

    Public Custom Event CustomHandler As EventHandler
        AddHandler(value As EventHandler)
            Caller.BeginInvoke("prop", Nothing, Nothing) ' Noncompliant
        End AddHandler
        RemoveHandler(value As EventHandler)
            Caller.BeginInvoke("prop", Nothing, Nothing) ' Noncompliant
        End RemoveHandler
        RaiseEvent(sender As Object, e As EventArgs)
            Caller.EndInvoke(Nothing)
        End RaiseEvent
    End Event

    Public Shared Widening Operator CType(F As ForCoverage) As Integer
        Caller.BeginInvoke("prop", Nothing, Nothing) ' Noncompliant
    End Operator

    Public Shared Operator +(a As ForCoverage, b As ForCoverage) As ForCoverage
        Caller.BeginInvoke("prop", Nothing, Nothing) ' Noncompliant
    End Operator

    Shared Sub New()
        Caller.BeginInvoke("Foo", Nothing, Nothing) ' Noncompliant
    End Sub

    Public Sub New()
        Caller.BeginInvoke("Foo", Nothing, Nothing) ' Noncompliant
    End Sub

    Public Sub Compliant()
        Dim Result As IAsyncResult = Caller.BeginInvoke("method", Nothing, Nothing) ' Compliant
        Caller.EndInvoke(Result)
    End Sub

    Public Sub Bar()
        Caller.BeginInvoke("Foo", Nothing, Nothing) ' Noncompliant
    End Sub

    Protected Overrides Sub Finalize()
        Caller.BeginInvoke("Foo", Nothing, Nothing) ' Noncompliant
    End Sub

    Public Sub BeginInvoke()
    End Sub

    Private Sub EndInvoke()
    End Sub

    Public Sub BeginInvoke(Arg As String)
    End Sub

    Private Sub InvokeFakeMethod()
        BeginInvoke()
        BeginInvoke("Wrong parameter type")
        EndInvoke()
    End Sub

    Private Sub InvokeSomethingElse()
        Dim BeginInvoke As String = "MemberBinding"
        Dim EndInvoke As String = "MemberBinding"
        BeginInvoke.ToString()
        EndInvoke.ToString()
    End Sub

    Private Sub InvokeFromInvocation()
        Caller.BeginInvoke("FromInvocation", CreateCallback(), Nothing) 'Noncompliant
    End Sub

    Private Function CreateCallback() As AsyncCallback
        Return Nothing
    End Function

    Public Structure FooStruct

        Public Field As String

        Public Sub New(Field As String)
            Me.Field = Field
            Caller.BeginInvoke("FooStruct", Nothing, Nothing) ' Noncompliant
        End Sub

    End Structure

End Class

' Dummy implementation of IAsyncResult compatible with both .Net Framework and .Net Core
Friend Class AsyncResult
    Implements IAsyncResult

    Public ReadOnly Property AsyncDelegate As Object

    Public ReadOnly Property IsCompleted As Boolean Implements IAsyncResult.IsCompleted
    Public ReadOnly Property AsyncWaitHandle As WaitHandle Implements IAsyncResult.AsyncWaitHandle
    Public ReadOnly Property AsyncState As Object Implements IAsyncResult.AsyncState
    Public ReadOnly Property CompletedSynchronously As Boolean Implements IAsyncResult.CompletedSynchronously

End Class

Class ReproEndinvokeDelegate

    Public Sub BeginInvokeWithEndinvokeDelegate()
        Dim A As Action = Sub() Return
        A.BeginInvoke(AddressOf A.EndInvoke, Nothing) ' Compliant
     End Sub

End Class
